File: PrimeMinistersByPython/csv2html/translator.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# import commands
import subprocess
import datetime
import locale
import os
import os.path
import re
import shutil

from PIL import Image

from .downloader import Downloader
from .table import Table
from .tuple import Tuple
from .writer import Writer
from .attributes import AttributesForPrimeMinisters
from .attributes import AttributesForTokugawaShogunate

class Translator(object):
	"""トランスレータ：CSVファイルをHTMLページへと変換するプログラム。"""

	def __init__(self, classOfAttributes):
		"""トランスレータのコンストラクタ。"""

		super(Translator, self).__init__()

		self._input_table = Table('input', classOfAttributes)
		self._output_table = Table('output', classOfAttributes)

		return

	def compute_string_of_days(self, period):
		"""在位日数を計算して、それを文字列にして応答する。"""

		period_list = re.split('〜', period)
		start = re.split('[年月日]', period_list[0])
		end = re.split('[年月日]', period_list[1])
		start_date = datetime.date(int(start[0]), int(start[1]), int(start[2]))
		if end[0] != '':
			end_date = datetime.date(int(end[0]), int(end[1]), int(end[2]))
		else:
			end_date = datetime.date.today()
		return str((end_date - start_date).days + 1)

	def compute_string_of_image(self, tuple):
		"""サムネイル画像から画像へ飛ぶためのHTML文字列を作成して、それを応答する。"""

		no_index = tuple.attributes().keys().index('no')
		image_index = tuple.attributes().keys().index('image')
		thumbnail_index = tuple.attributes().keys().index('thumbnail')
		img_html = '<a name="{no}" href="{image}"><img class="borderless" src="{thumbnail}" width="25" height="32" alt="{alt}"></a>'.format(
			no = tuple.values()[no_index],
			image = tuple.values()[image_index],
			thumbnail = tuple.values()[thumbnail_index],
			alt = tuple.values()[image_index].split('/')[1],
		)
		return img_html

	def execute(self):
		"""CSVファイルをHTMLページへと変換する。"""

		# ダウンローダに必要なファイル群をすべてダウンロードしてもらい、
		# 入力となるテーブルを獲得する。
		a_downloader = Downloader(self._input_table)
		a_downloader.perform()

		# トランスレータに入力となるテーブルを渡して変換してもらい、
		# 出力となるテーブルを獲得する。
		print(self._input_table)
		self.translate()
		print(self._output_table)

		# ライタに出力となるテーブルを渡して、
		# Webページを作成してもらう。
		a_writer = Writer(self._output_table)
		a_writer.perform()

		# 作成したページをウェブブラウザで閲覧する。
		class_attributes = self._output_table.attributes().__class__
		base_directory = class_attributes.base_directory()
		index_html = class_attributes.index_html()
		a_command = 'open -a Safari ' + base_directory + os.sep + index_html
		subprocess.getoutput(a_command)

		return

	@classmethod
	def perform(the_class, class_attributes):
		"""属性リストのクラスを受け取り、CSVファイルをHTMLページへと変換する。"""

		# トランスレータのインスタンスを生成する。
		a_translator = the_class(class_attributes)
		# トランスレータにCSVファイルをHTMLページへ変換するように依頼する。
		a_translator.execute()

		return

	def translate(self):
		"""CSVファイルを基にしたテーブルから、HTMLページを基にするテーブルに変換する。"""

		for tuple in self._input_table.tuples():
			values = []
			for attribute in self._output_table.attributes().keys():
				if attribute == "image":
					values.append(self.compute_string_of_image(tuple))
				elif attribute == "days":
					period_index_in_tuple = tuple.attributes().keys().index("period")
					values.append(self.compute_string_of_days(tuple.values()[period_index_in_tuple]))
				elif attribute in self._input_table.attributes().keys():
					attribute_index_in_tuple = tuple.attributes().keys().index(attribute)
					values.append(tuple.values()[attribute_index_in_tuple])
				else:
					values.append(None)
			self._output_table.add(Tuple(self._output_table.attributes(), values))
		return

File: PrimeMinistersByJava/csv2html/Translator.java
package csv2html;

import java.awt.image.BufferedImage;
import java.io.File;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashSet;
import java.util.List;

/**
* トランスレータ：CSVファイルをHTMLページへと変換するプログラム。
*/
public class Translator extends Object
{
	/**
	* CSVに由来するテーブルを記憶するフィールド。
	*/
	private Table inputTable;

	/**
	* HTMLに由来するテーブルを記憶するフィールド。
	*/
	private Table outputTable;

	/**
	* 属性リストのクラスを指定したトランスレータのコンストラクタ。
	* @param classOfAttributes 属性リストのクラス
	*/
	public Translator(Class<? extends Attributes> classOfAttributes)
	{
		super();

		Attributes.flushBaseDirectory();

		try
		{
			Constructor<? extends Attributes> aConstructor = classOfAttributes.getConstructor(String.class);

			this.inputTable = new Table(aConstructor.newInstance("input"));
			this.outputTable = new Table(aConstructor.newInstance("output"));
		}
		catch (NoSuchMethodException |
		InstantiationException |
		IllegalAccessException |
		InvocationTargetException anException) { anException.printStackTrace(); }

		return;
	}

	/**
	* 在位日数を計算して、それを文字列にして応答する。
	* @param periodString 在位期間の文字列
	* @return 在位日数の文字列
	*/
	public String computeNumberOfDays(String periodString)
	{
		List<String> aPeriodList = Arrays.asList(periodString.split("[〜]"));
		List<String> aStartParams = Arrays.asList(aPeriodList.get(0).split("[年月日]"));
		int startYear = Integer.parseInt(aStartParams.get(0));
		int startMonth = Integer.parseInt(aStartParams.get(1)) - 1;
		int startDay = Integer.parseInt(aStartParams.get(2));
		Calendar aStartCalender = Calendar.getInstance();
		Calendar aEndCalender = Calendar.getInstance();
		aStartCalender.set(startYear, startMonth, startDay);
		if (aPeriodList.size() >= 2)
		{
			List<String> aEndParams = Arrays.asList(aPeriodList.get(1).split("[年月日]"));
			int endYear = Integer.parseInt(aEndParams.get(0));
			int endMonth = Integer.parseInt(aEndParams.get(1)) - 1;
			int endDay = Integer.parseInt(aEndParams.get(2));
			aEndCalender.set(endYear, endMonth, endDay);
		}
		long diffTimeOfMilliseconds = aEndCalender.getTimeInMillis() - aStartCalender.getTimeInMillis();
		long dayTime = (diffTimeOfMilliseconds / (1000*60*60*24)) + 1;
		String days = String.format("%1$d" ,dayTime);
		return days;
	}

	/**
	* サムネイル画像から画像へ飛ぶためのHTML文字列を作成して、それを応答する。
	* @param aString 画像の文字列
	* @param aTuple タプル
	* @param no 番号
	* @return サムネイル画像から画像へ飛ぶためのHTML文字列
	*/
	public String computeStringOfImage(String aString, Tuple aTuple, int no)
	{
		BufferedImage image = this.inputTable.thumbnails().get(no);
		List<String> anyStrings = aTuple.values();
		String index = anyStrings.get(aTuple.attributes().indexOfNo());
		String imageString = anyStrings.get(aTuple.attributes().indexOfImage());
		String thumnailString = anyStrings.get(aTuple.attributes().indexOfThumbnail());
		StringBuilder aBuilder = new StringBuilder();
		aBuilder.append("<a name=\"");
		aBuilder.append(thumnailString);
		aBuilder.append("\"href=\"");
		aBuilder.append(aString);
		aBuilder.append("\">");
		aBuilder.append("<img class=\"borderless\" src=\"");
		aBuilder.append(thumnailString);
		aBuilder.append("\" width=\"");
		aBuilder.append(image.getWidth());
		aBuilder.append("\" height=\"");
		aBuilder.append(image.getHeight());
		aBuilder.append("\" alt=\"");
		aBuilder.append(imageString.split("/")[1]);
		aBuilder.append("\"></a>");

		return aBuilder.toString();
	}

	/**
	* CSVファイルをHTMLページへ変換する。
	*/
	public void execute()
	{
		// 必要な情報をダウンロードする。
		Downloader aDownloader = new Downloader(this.inputTable);
		aDownloader.perform();

		// CSVに由来するテーブルをHTMLに由来するテーブルへと変換する。
		System.out.println(this.inputTable);
		this.translate();
		System.out.println(this.outputTable);

		// HTMLに由来するテーブルから書き出す。
		Writer aWriter = new Writer(this.outputTable);
		aWriter.perform();

		// ブラウザを立ち上げて閲覧する。
		try
		{
			Attributes attributes = this.outputTable.attributes();
			String fileStringOfHTML = attributes.baseDirectory() + attributes.indexHTML();
			ProcessBuilder aProcessBuilder = new ProcessBuilder("open", "-a", "Safari", fileStringOfHTML);
			aProcessBuilder.start();
		}
		catch (Exception anException) { anException.printStackTrace(); }

		return;
	}

	/**
	* 属性リストのクラスを受け取って、CSVファイルをHTMLページへと変換するクラスメソッド。
	* @param classOfAttributes 属性リストのクラス
	*/
	public static void perform(Class<? extends Attributes> classOfAttributes)
	{
		// トランスレータのインスタンスを生成する。
		Translator aTranslator = new Translator(classOfAttributes);
		// トランスレータにCSVファイルをHTMLページへ変換するように依頼する。
		aTranslator.execute();

		return;
	}

	/**
	* CSVファイルを基にしたテーブルから、HTMLページを基にするテーブルに変換する。
	*/
	public void translate()
	{
		int index = 0;
		int no = 0;
		int indexOfThumbnail = this.inputTable.attributes().indexOfThumbnail();
		List<String> listCollection = new ArrayList<String>();
		for(String aString : this.inputTable.attributes().names())
		{
			if(index != indexOfThumbnail)
			{
				listCollection.add(IO.htmlCanonicalString(aString));
				if(index == this.inputTable.attributes().indexOfPeriod()){ listCollection.add(IO.htmlCanonicalString("在位日数"));	}
			}
			index++;
		}
		this.outputTable.attributes().names(listCollection);
		for(Tuple aTuple : this.inputTable.tuples())
		{
			index = 0;
			List<String> aCollection = new ArrayList<String>();
			for(String aAttribute : this.outputTable.attributes().keys())
			{
				if(index == this.outputTable.attributes().indexOfImage())
				{
					Integer imageIndexInTuple = aTuple.attributes().indexOfImage();
					aCollection.add(this.computeStringOfImage(aTuple.values().get(imageIndexInTuple),aTuple,index));
				}
				else if(index == this.outputTable.attributes().indexOfDays())
				{
					Integer periodIndexInTuple = aTuple.attributes().indexOfPeriod();
					aCollection.add(this.computeNumberOfDays(aTuple.values().get(periodIndexInTuple)));
				}
				else if(this.inputTable.attributes().indexOf(aAttribute) != -1)
				{
					Integer attributeIndexInTuple = aTuple.attributes().indexOf(aAttribute);
					aCollection.add(aTuple.values().get(attributeIndexInTuple));
				}
				else
				{
					aCollection.add("");
				}
				index++;
			}
			Tuple outTuple = new Tuple(this.outputTable.attributes(), aCollection);
			this.outputTable.add(outTuple);
			no++;
		}
		return;
	}
}
